lib/wallet/src/Cardano/Wallet/IO/Delegation.hs

```diff
@@ -528,3 +494,109 @@ quitStakePool ctx walletId passphrase = do
     netLayer = ctx ^. networkLayer
     ti = timeInterpreter netLayer
     txLayer = ctx ^. transactionLayer
+
+{-----------------------------------------------------------------------------
+    Join DRep
+------------------------------------------------------------------------------}
+-- | Send a transaction to the network where we join drep.
+joinDRep
+    :: forall s n k.
+        ( s ~ SeqState n k
+        , WalletFlavor s
+        , Excluding '[SharedKey] k
+        , AddressIndexDerivationType k ~ 'Soft
+        , GenChange s
+        , IsOurs (SeqState n k) RewardAccount
+        , AddressBookIso s
+        , Seq.SupportsDiscovery n k
+        , DelegationAddress k 'CredFromKeyK
+        )
+    => WalletLayer IO s
+    -> WalletId
+    -> DRep
+    -> Passphrase "user"
+    -> IO (W.BuiltTx, UTCTime)
+joinDRep ctx wid drep passphrase = do
+    pp <- currentProtocolParameters netLayer
+    ttl <- W.transactionExpirySlot ti Nothing
+
+    voting <- handleVoteRequest ctx drep
+
+    let transactionCtx =
+            defaultTransactionCtx
+                { txWithdrawal = NoWithdrawal
+                , txValidityInterval = (Nothing, ttl)
+                , txDelegationAction = Nothing
+                , txVotingAction = Just voting
+                , txDeposit = Just $ stakeKeyDeposit pp
+                }
+
+    let changeAddrGen = W.defaultChangeAddressGen @s (delegationAddressS @n)
+    W.buildSignSubmitTransaction @s
+        db
+        netLayer
+        txLayer
+        passphrase
+        wid
+        changeAddrGen
+        (PreSelection [])
+        transactionCtx
+  where
+    db = ctx ^. dbLayer
+    netLayer = ctx ^. networkLayer
+    ti = timeInterpreter netLayer
+    txLayer = ctx ^. transactionLayer
+
+handleVoteRequest
+    :: WalletLayer IO s
+    -> DRep
+    -> IO Tx.VotingAction
+handleVoteRequest ctx drep = do
+    (vAction, votingRequest) <- voteAction ctx drep
+    (Write.PParamsInAnyRecentEra era _, _)
+        <- W.readNodeTipStateForTxWrite netLayer
+
+    either (throwIO . ExceptionVoting) pure
+        (WD.guardOnlyVoting era $ toDrepEnriched votingRequest)
+    pure vAction
+  where
+    toDrepEnriched NotVotedYet = Nothing
+    toDrepEnriched VotedSameAsBefore = Just (True, drep)
+    toDrepEnriched VotedDifferently = Just (False, drep)
+
+    netLayer = ctx ^. networkLayer
```

Comment:
This function is okay, but harder to understand than necessary — in particular, it's not clear what exactly `toDrepEnrichted` is supposed to be doing. Could you remove this function (`handleVoteRequest`) and structure the function `joinDRep` similar to the function `quitStakePool` instead?

In other words, I would like you to structure the code in the following way:

```hs
joinDRep — IO, collects relevant parameter values, calls

    joinDRepVotingAction — IO, provides additional parameter values (if necessary), calls

        WD.joinDRepVotingAction — *pure*, computes a result based on all parameter values.
```

The idea is that the functions `joinDRep` and `joinDRepVotingAction` are responsible for providing sufficient parameter values to compute the result, but the actual computation of the result (= the `Either ErrCannotJoin Tx.VotingAction` is contained within the pure function. In other words, the concern of figuring out what the vote action should be — if any — should be entirely confined to the definition of the pure function `WD.joinDRepVotingAction`.

In contrast, at the moment, this function uses `toDRepEnriched` to package some intermediate computation and hands it off to the function `WD.guardOnlyVoting ` in a different module, but the meaning of the intermediate data format `Maybe (Bool, DRep)` is not apparent from the type signatures and function names. The logic becomes much clear if:

* Use pure functions to compute the result.
* Keep the computation of the result to a single module

Could you change this?

Reply:
sounds reasonable! done